Automattic/wp-calypso
# Signup Flow: Fix paid items missing from cart by removing siteSlug dependency when changing flows (when it makes sense to).

## The problem

### The problem from the user perspective:

When a user attempts to:
- launch a site (via `/start/launch-site`) 
  - but abandons halfway
- then goes on to create a new site (via `/start`) with a paid plan
  - the paid plan goes missing from the cart.

### The problem from the technical perspective:

On the `/start` flow, this is the normal order of operation:
- `createSiteWithCart` is called first
- `addPlanToCart` is called after

When the `/start/launch-site` flow is abandoned halfway, the `siteSlug` persists in the data store.

When this `siteSlug` exists, the order of operation on `/start` executes in reverse:
- `addPlanToCart` ended up getting called first
- `createSiteWithCart` gets called after.

The order of operation becomes reversed because (see [steps-pure.js](https://github.com/Automattic/wp-calypso/blob/trunk/client/signup/config/steps-pure.js) for reference):
- the `domains` step has the option `delayApiRequestUntilComplete: true` for `createSiteWithCart`
   - this means that other steps can go ahead and execute their api requests first (if their dependency is fulfilled).
- the `plans` step has the `siteSlug` dependency fulfilled (because it was persisted through previous flow).
   - this resulted in `addPlanToCart` api request being called first, then only `createSiteWithCart` api request kicks in.

## Changes proposed in this Pull Request

### The fix

The fix is quite simply to remove the `siteSlug` when transitioning from a **launch** flow to a **onboarding** flow.

### The challenge

However, [there are many flows](https://github.com/Automattic/wp-calypso/blob/trunk/client/signup/config/flows-pure.js), how do you determine what is a **launch** flow and what is an **onboarding** flow?

### The proposed solution

Instead of determining what flow it is, we determine whether removing the `siteSlug` is necessary when:
- the previous flow **does not have a step** that provides the `siteSlug`
- the current flow **has a step** that provides the `siteSlug`

The reasoning behind this is that:
- if the **current flow** has a step that provides `siteSlug`
  - we should rely on that step in that flow to provide the new `siteSlug`
- instead of relying on the `siteSlug` that was persisted by the **previous flow**.

## Testing instructions

Repeat the steps below:
- once for `wordpress.com` to replicate the failure
- once for `calypso.localhost:3000` to see if the failure is fixed

#### Create an unlaunched site

1. Clear all site data via **Dev Tool > Application > Storage > Clear site data**.
2. Login and create a new site via `/start`.

#### Enter the launch flow for the unlaunched site

1. Go to **My Home > Site Setup > Launch your site > Launch site**.
2. When you're in the launch flow, the url could be either `/start/launch-site` or `/start/new-launch`.
3. Go to **Dev Tool > Application > Storage > IndexedDB** under **calypso_store > redux-state-123456:signup**:
   - You should see your `siteSlug` under the `dependencyStore` object.

![image](https://user-images.githubusercontent.com/1287077/116687075-49fa8500-a9bd-11eb-9b32-9a2dd368b750.png)

#### Abandon the launch flow and go to onboarding flow.

1. Open a new tab on your browser.
   - If you're retesting again & again, do not take shortcuts. If you already have a tab with `/start` with open, do not refresh that tab, this will only cause the `previousFlowName` to be set to `onboarding`, and then you'll have to repeat all the process above again.
2. Go to `/start`.
3. Go to **Dev Tool > Application > Storage > IndexedDB** under **calypso_store > redux-state-123456:signup** again:
   - For `wordpress.com`, you should see your `siteSlug` under the `dependencyStore` object.
   - For `calypso.localhost:3000`, you should see:
     - no `siteSlug` value under the `dependencyStore` object 
     - the `currentFlowName: "onboarding"` value and `previousFlowName: "launch-site"` value under the `flow` object.
     
![image](https://user-images.githubusercontent.com/1287077/116686509-9396a000-a9bc-11eb-8944-3fa0e264355d.png)


#### Test the checkout step
1. Try to create a site with a paid plan.
2. At the checkout screen:
   - For `wordpress.com`, you will encounter the error **The cart is empty.**
   - For `calypso.localhost:3000`, you will see the added plan in your cart.

Related to #50430 https://github.com/Automattic/wp-calypso/pull/51021

Might be also related to:
- https://github.com/Automattic/wp-calypso/issues/35363
- https://github.com/Automattic/wp-calypso/issues/28196
